### PR TITLE
core: move gas verification to block

### PR DESCRIPTION
**Pull request by forcodedancing**
### Description

This pr will move `gas account` merkle proof verification from tx level to block level, which will reduce the number of constraints a lot.

### Rationale

For most l2 transactions, only one gas asset will be involved in a tx. However, for `AtomicMatch`, two kinds of assets will be send to `gas account`. Therefore, we use a slice of assets for recording gas assets.

### Example

NA

### Changes

Notable changes:
* reduce merke proof accounts

### CODE REVIEW

File: circuit/types/constants.go [hunk 1]
```diff
@@ -39,7 +39,12 @@ const (
 	DefaultInt = int64(-1)
 
 	NbAccountAssetsPerAccount = 4
-	NbAccountsPerTx           = 5
+	NbAccountsPerTx           = 4
+
+	GasAccountIndex = 1 // gas account index
```
Comment (lightning-li) [inline circuit/types/constants.go:44]: GasAccountIndex should be a variable rather than const

File: circuit/gas_constraints.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	)
+	accountNodeHash = hFunc.Sum()
+	hFunc.Reset()
+	// update merkle proof
+	NewAccountRoot = types.UpdateMerkleProof(api, hFunc, accountNodeHash, gas.MerkleProofsAccountBefore[:], accountIndexMerkleHelper)
+
+	hFunc.Reset()
+	hFunc.Write(
+		NewAccountRoot,
+		lastRoots[0],
```
Comment (lightning-li) [inline circuit/gas_constraints.go:122]: This should be lastRoots[1], and next line should be lastRoots[2]?

File: circuit/types/constants.go (hunk 1 above)
Comment (forcodedancing) [reply circuit/types/constants.go:44]: sure

File: circuit/gas_constraints.go (hunk 2 above)
Comment (forcodedancing) [reply circuit/gas_constraints.go:122]: yes, typos

**Pull request closed without merging**